src/cpager/cpager.c

```diff
@@ -1,35 +1,84 @@
+#include <criu.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <fcntl.h>
 #include <Python.h>
 
+#define RAISE_PYTHON_ERROR_ON_FAIL(arg) if (arg < 0) { PyErr_Format(PyExc_ValueError, "%s (%d)", strerror(errno), arg); return NULL; }
+#define LOG_FILE "log.txt"
 
 
 PyObject *module = NULL;
 
 
-static PyObject* run_on_line(PyObject* self) {
-    // call whatever you want here
-    printf("Call whatever you want here...\n");
-    fflush(stdout);
+static PyObject* take_snapshot(PyObject* self, PyObject *args) {
+    int pid = 0;
+    int dir_fd = 0;
+    int res = 0;
+    const char* file_path = NULL;
+
+    if (!PyArg_ParseTuple(args, "is", &pid, &file_path))  {
+        return NULL;
+    }
+
+    dir_fd = open(file_path, O_DIRECTORY);
+
+    criu_init_opts();
+    criu_set_pid(pid);
+    criu_set_tcp_established(true);
+    criu_set_log_file(LOG_FILE);
+    criu_set_shell_job(true);
+    criu_set_leave_running(true);
+    criu_set_work_dir_fd(dir_fd);
+    criu_set_images_dir_fd(dir_fd);
+    RAISE_PYTHON_ERROR_ON_FAIL(criu_dump());
+
     return Py_BuildValue("");
-}
+ }
 
 
-static PyMethodDef methods[]= {
-    {"run_on_line", (PyCFunction)run_on_line, METH_NOARGS, "runs on each line"},
-    {NULL, NULL, NULL, 0, NULL}
-};
+static PyObject* restore_from_snapshot(PyObject* self, PyObject* args) {
+    int dir_fd = 0;
+    char* file_path = NULL;
 
+    if (!PyArg_ParseTuple(args, "s", &file_path))  {
+        return NULL;
+    }
 
-static struct PyModuleDef cpagerModule = {
-    PyModuleDef_HEAD_INIT,
-    "cpager",
-    "description",
-    -1,
-    methods
-};
+    dir_fd = open(file_path, O_DIRECTORY);
 
+    criu_init_opts();
```

Comment:
What if any of these fail?

Reply:
They all return void, only the last one is checkable